specs/Standards/MultiToken/README.md

```diff
@@ -0,0 +1,539 @@
+pub trait MultiTokenResolver {
+    /// Finalize an `mt_transfer_call` chain of cross-contract calls.
+    ///
+    /// The `mt_transfer_call` process:
+    ///
+    /// 1. Sender calls `mt_transfer_call` on MultiToken contract
+    /// 2. MultiToken contract transfers token from sender to receiver
+    /// 3. MultiToken contract calls `mt_on_transfer` on receiver contract
+    /// 4+. [receiver contract may make other cross-contract calls]
+    /// N. MultiToken contract resolves promise chain with `mt_resolve_transfer`, and may
+    ///    transfer token back to sender
+    ///
+    /// Requirements:
+    /// * Contract MUST forbid calls to this function by any account except self
+    /// * If promise chain failed, contract MUST revert token transfer
+    /// * If promise chain resolves with `true`, contract MUST return token to
+    ///   `sender_id`
+    ///
+    /// Arguments:
+    /// * `previous_owner_id`: the owner prior to the call to `mt_transfer_call`
+    /// * `receiver_id`: the `receiver_id` argument given to `mt_transfer_call`
+    /// * `token_ids`: the `token_ids` argument given to `mt_transfer_call`
+    /// * `approvals`: if using Approval Management, contract MUST provide
+    ///   set of original approved accounts in this argument, and restore these
+    ///   approved accounts in case of revert. In this case it may be multiple sets of approvals
+    ///
+    /// Returns true if tokens were successfully transferred to `receiver_id`.
+    fn mt_resolve_transfer(
+        &mut self,
+        sender_id: AccountId,
+        receiver_id: AccountId,
+        token_ids: Vec<TokenId>,
+        amounts: Vec<U128>,
+    ) -> Vec<U128>;
+}
+```
+### Storage Management Trait
+#### Notes
+This is semi necessary for ft token types to be able to refund users for storage of many different token types like gold/silver... this might be slightly out of scope
+```
+pub trait StorageManagement {
+    // if `registration_only=true` MUST refund above the minimum balance if the account didn't exist and
+    //     refund full deposit if the account exists.
+    fn storage_deposit(
+        &mut self,
+        token_ids: Vec<TokenId>,
+        account_id: Option<AccountId>,
+        registration_only: Option<bool>,
+    ) -> StorageBalance;
+
+    /// Withdraw specified amount of available Ⓝ for predecessor account.
+    ///
+    /// This method is safe to call. It MUST NOT remove data.
+    ///
+    /// `amount` is sent as a string representing an unsigned 128-bit integer. If
+    /// omitted, contract MUST refund full `available` balance. If `amount` exceeds
+    /// predecessor account's available balance, contract MUST panic.
+    ///
+    /// If predecessor account not registered, contract MUST panic.
+    ///
+    /// MUST require exactly 1 yoctoNEAR attached balance to prevent restricted
+    /// function-call access-key call (UX wallet security)
+    ///
+    /// Returns the StorageBalance structure showing updated balances.
```

Comment:
Is the `StorageBalance` structure still accurate/correct for this standard?  Do we need to be able to keep track of the amount of storage _each token type_ requires?

Reply:
Here's an example of what this looks like in practice. https://github.com/shipsgold/multi-token-standard-impl/blob/feat/initial-token/multi_token/src/core/storage_impl.rs
 
 In this standard there is no explicit token type, that information is encapsulated by token id , where developers can decide how they'd like to label or identify the type of token they'd like to have. This is mostly in alignment with erc-1155. The difference is erc-1155 restricts tokenId to uint256 so users must bitshift and figure out what kind of mask if any mask they want to put on token id. This same thing can be done with string if users care to implement it.
 
 https://github.com/shipsgold/multi-token-standard-impl/blob/feat/initial-token/

Reply:
also from the meeting notes https://github.com/shipsgold/multi-token-standard-impl/blob/main/meetings/2021-08-17-meeting-notes.md#storage-management-overview-2600 for background.